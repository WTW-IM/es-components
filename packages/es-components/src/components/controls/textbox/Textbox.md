The Textbox component will accept typical input attributes as props such as onChange, onBlur, value, placeholder, etc.

```
import Control from '../Control';
import Label from '../label/Label';

<>
  <Control>
    <Label htmlFor="stacked">Stacked</Label>
    <Textbox id="stacked" className="custom-class" />
  </Control>

  <Control orientation="inline">
    <Label htmlFor="inline">Inline</Label>
    <Textbox id="inline" />
  </Control>
</>
```

Pass an `onChange` function to execute any time the input box value changes. This function will have the current value of the input box passed to it. The following is a simple controlled example, where state is handled outside the component. (See [Controlled Components](https://facebook.github.io/react/docs/htmlForms.html#controlled-components) and [Uncontrolled Components](https://facebook.github.io/react/docs/uncontrolled-components.html).)

```
import Control from '../Control';
import Label from '../label/Label';

function TextboxExample() {
  const [value, setValue] = React.useState('')

  function handleOnTextChange(event) {
    console.log(event.target.value)
    setValue(event.target.value)
  }

  return (
    <Control>
      <Label htmlFor="controlled-example">Controlled Example</Label>
      <Textbox
        id="controlled-example"
        value={value}
        onChange={handleOnTextChange}
      />
    </Control>
  )
};

<TextboxExample />
```

Pass an `onBlur` function to execute when the text box loses focus. This function will have the current value of the input passed to it. To provide a default value in an uncontrolled `Textbox` component, use `defaultValue`.

```
import Control from '../Control';
import Label from '../label/Label';

function handleOnBlur(event) {
  if(event.target.value.length > 0) {
    confirm(`You entered ${event.target.value}. Is that correct?`);
  }
}

<Control>
  <Label htmlFor="first-name">First name</Label>
  <Textbox
    id="first-name"
    defaultValue="Default"
    onBlur={handleOnBlur}
  />
</Control>
```

### Validation states

Validation states are driven from the `Control` component.

<a href="https://8lf1uv.axshare.com/#id=mjvdz7&p=incomplete_fields&dp=0&g=1" target="blank"><div style="color:#cc0000;text-transform:uppercase;margin-bottom:1em;">View Error Pattern Examples</div></a>

```
import Control from '../Control';
import Label from '../label/Label';
import AdditionalHelp from '../AdditionalHelp';

<>
  <Control validationState="success">
    <Label htmlFor="success">Success</Label>
    <Textbox id="success" aria-describedby="success-help" />
    <AdditionalHelp id="success-help">When validationState is set to success</AdditionalHelp>
  </Control>

  <Control validationState="warning">
    <Label htmlFor="warning">Warning</Label>
    <Textbox id="warning" aria-describedby="warning-help" />
    <AdditionalHelp id="warning-help">When validationState is set to warning</AdditionalHelp>
  </Control>

  <Control validationState="danger">
    <Label htmlFor="danger">Danger</Label>
    <Textbox id="danger" aria-describedby="danger-help" />
    <AdditionalHelp id="danger-help">When validationState is set to danger</AdditionalHelp>
  </Control>
</>
```

### Appending and Prepending text

Provide an `appendText` or `prependText` prop htmlFor appending and prepending inputs. Validation state colorings will also be applied.

```
import Control from '../Control';
import Label from '../label/Label';

<>
  <Control>
    <Label htmlFor="appended">Appended</Label>
    <Textbox
      id="appended"
      appendIconName="calendar"
    />
  </Control>

  <Control>
    <Label htmlFor="prepended">Prepended</Label>
    <Textbox
      id="prepended"
      prependIconName="phone"
    />
  </Control>

  <Control>
    <Label htmlFor="both">Appended and prepended</Label>
    <Textbox
      id="both"
      appendIconName="user"
      prependIconName="phone"
    />
  </Control>

  <Control validationState="success">
    <Label htmlFor="state-appended">Appended</Label>
    <Textbox
      id="state-appended"
      appendIconName="calendar"
    />
  </Control>

  <Control validationState="warning">
    <Label htmlFor="state-prepended">Prepended</Label>
    <Textbox
      id="state-prepended"
      prependIconName="phone"
    />
  </Control>

  <Control validationState="danger">
    <Label htmlFor="state-both">Appended and prepended</Label>
    <Textbox
      id="state-both"
      appendIconName="user"
      prependIconName="phone"
    />
  </Control>
</>
```

### Additional props

Other typical input attributes passed to `Textbox` are allowed. The react `autoFocus` property is also accepted.

```
import Control from '../Control';
import Label from '../label/Label';

<>
  <Control>
    <Label htmlFor="disabled">Disabled</Label>
    <Textbox
      id="disabled"
      defaultValue="John Doe"
      disabled
    />
  </Control>

  <Control>
    <Label htmlFor="readonly">Readonly</Label>
    <Textbox
      id="readonly"
      defaultValue="Jane Doe"
      readOnly
    />
  </Control>

  <Control>
    <Label htmlFor="placeholder">Placeholder</Label>
    <Textbox
      id="placeholder"
      placeholder="Placeholder is one of many input attributes"
    />
  </Control>
</>
```
