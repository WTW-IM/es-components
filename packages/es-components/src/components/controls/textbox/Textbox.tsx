import React, {
  useCallback,
  useImperativeHandle,
  useState,
  useEffect,
  useMemo
} from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { IconName, iconNames } from 'es-components-shared-types';
import getStyledProp, { ESThemeProps } from '../../util/getStyledProp';
import Icon, { iconBaseStyles, IconBaseProps } from '../../base/icons/Icon';
import InputBase, {
  useValidationStyleProps,
  validationStateHighlightStyles,
  validationStateInputStyles,
  validationStateSetupStyles,
  noInset,
  basicTextboxStyles,
  propTypes as basePropTypes,
  defaultProps as baseDefaultProps,
  ValidationStyleProps
} from './InputBase';
import { callRefs } from '../../util/callRef';

type GetComputedStyleFunc = typeof window.getComputedStyle;
type ComputedStyleParams = Parameters<GetComputedStyleFunc>;
type ComputedStyleReturn = ReturnType<GetComputedStyleFunc>;
type IconElement = HTMLElement;

const computeStyle = (...args: ComputedStyleParams) => {
  try {
    return window.getComputedStyle(...args);
  } catch (err) {
    if ((err as Error)?.message?.match(/Not Implemented/i)) {
      return {} as ComputedStyleReturn;
    }
    throw err;
  }
};

type IconStyleProps = ValidationStyleProps & IconBaseProps;

const getIconStyles = (icon: IconElement) => {
  const computedStyle = computeStyle(icon);
  const beforeStyle = computeStyle(icon, ':before');
  return css<IconStyleProps>`
    ${props => css`
      ${iconBaseStyles}

      // computed styles
      content: ${beforeStyle.content};
      font-family: ${computedStyle.fontFamily};
      font-size: ${computedStyle.fontSize};

      // themed addon styles
      background-color: ${getStyledProp('backgroundColor', 'addOn', props)};
      color: ${getStyledProp('textColor', 'addOn', props)};

      // base addon styles
      box-sizing: border-box;
      display: flex;
      flex-direction: column;
      justify-content: center;
      line-height: 1;
      margin: 0;
      outline: 0;
      padding: 0.333em 0.6111em;
    `}
  `;
};

const useIconStyles = (
  iconName: IconName | undefined,
  icon: Maybe<IconElement>
) => {
  const [iconRecheck, setIconRecheck] = useState(0);
  const iconStyles = useMemo(
    () => (icon ? getIconStyles(icon) : css``),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [icon, iconRecheck]
  );

  useEffect(
    function ensureIconRendered() {
      if (!iconName || icon) {
        return;
      }

      requestAnimationFrame(() => setIconRecheck(r => r + 1));
    },
    [iconName, icon]
  );

  return iconStyles;
};

type IconStyles = ReturnType<typeof useIconStyles>;

export type TextboxAdditionProps = {
  hasPrepend?: boolean;
  hasAppend?: boolean;
};

export type InputWrapperProps = IconStyleProps &
  TextboxAdditionProps & {
    prependIconStyles: IconStyles;
    appendIconStyles: IconStyles;
  };

const InputWrapper = styled.div<InputWrapperProps>`
  ${validationStateSetupStyles}
  ${validationStateInputStyles}
  display: flex;

  padding: 0 !important;

  &:focus-within {
    ${validationStateHighlightStyles}
  }

  ${({ prependIconStyles }) =>
    css`
      &&:before {
        ${prependIconStyles}
      }
    `}

  ${({ appendIconStyles }) =>
    css`
      &&:after {
        ${appendIconStyles}
      }
    `}
`;

export const TextboxBase = styled(InputBase)<TextboxAdditionProps>`
  ${basicTextboxStyles}
  width: 100%;

  && {
    border: none !important;
    margin: 0 !important;
    height: ${props =>
      getStyledProp('inputStyles.inputHeight', props as ESThemeProps) ||
      '2.2em'};
  }

  ${({ hasPrepend }) =>
    hasPrepend &&
    css`
      border-bottom-left-radius: 0;
      border-top-left-radius: 0;
    `}

  ${({ hasAppend }) =>
    hasAppend &&
    css`
      border-bottom-right-radius: 0;
      border-top-right-radius: 0;
    `}

  &&:focus {
    box-shadow: ${noInset};
  }
`;

const Hidden = (props: JSXElementProps<'div'>) => (
  <div
    aria-hidden="true"
    style={{
      position: 'absolute',
      visibility: 'hidden',
      overflow: 'visible',
      width: 0,
      height: 0,
      left: -5000
    }}
    {...props}
  />
);

export interface TextboxProps extends JSXElementProps<'input'> {
  flat?: boolean;
  prependIconName?: IconName;
  appendIconName?: IconName;
}

const Textbox = React.forwardRef<Maybe<HTMLInputElement>, TextboxProps>(
  function ForwardedTextbox(props, ref) {
    const {
      prependIconName,
      appendIconName,
      type = 'text',
      flat,
      className,
      style,
      ...additionalTextProps
    } = props;
    const validationProps = useValidationStyleProps({ flat });

    const [inputRef, setInputRef] = useState<Maybe<HTMLInputElement>>();

    const [prependIcon, setPrependIcon] = useState<Maybe<HTMLElement>>();
    const [appendIcon, setAppendIcon] = useState<Maybe<HTMLElement>>();
    const prependIconStyles = useIconStyles(prependIconName, prependIcon);
    const appendIconStyles = useIconStyles(appendIconName, appendIcon);

    useImperativeHandle(ref, () => inputRef || undefined, [inputRef]);

    const setRefs = useCallback(
      (instance: Maybe<HTMLInputElement>) => {
        callRefs(instance, setInputRef, ref);
      },
      [ref]
    );

    const hasPrepend = !!prependIconName;
    const hasAppend = !!appendIconName;
    const prependProps = { hasPrepend, hasAppend };
    const sharedProps = { ...validationProps, ...prependProps };

    const focusInput = useCallback(() => inputRef?.focus(), [inputRef]);

    return (
      <InputWrapper
        className={className}
        as="div"
        onClick={focusInput}
        style={style}
        {...{
          ...sharedProps,
          prependIconStyles,
          appendIconStyles
        }}
      >
        {hasPrepend || hasAppend ? (
          <Hidden>
            <Icon ref={setPrependIcon} name={prependIconName} size={18} />
            <Icon ref={setAppendIcon} name={appendIconName} size={18} />
          </Hidden>
        ) : (
          <></>
        )}
        <TextboxBase
          ref={setRefs}
          type={type}
          className={className}
          {...additionalTextProps}
          {...sharedProps}
        />
      </InputWrapper>
    );
  }
);

const inputTypes = [
  'button',
  'checkbox',
  'color',
  'date',
  'datetime-local',
  'email',
  'file',
  'hidden',
  'image',
  'month',
  'number',
  'password',
  'radio',
  'range',
  'reset',
  'search',
  'submit',
  'tel',
  'text',
  'time',
  'url',
  'week'
];

type InputTypeAttribute = JSXElementProps<'input'>['type'];

export const propTypes = {
  ...basePropTypes,
  /** Content to prepend input box with */
  prependIconName: PropTypes.oneOf<IconName>(iconNames),
  /** Content to append to input box */
  appendIconName: PropTypes.oneOf<IconName>(iconNames),
  /** The type attribute for the textbox input */
  type: PropTypes.oneOf<InputTypeAttribute>(inputTypes),
  /** Whether the textbox is the flat style or not */
  flat: PropTypes.bool
};

export const defaultProps = {
  ...baseDefaultProps,
  prependIconName: undefined,
  appendIconName: undefined,
  type: 'text',
  flat: undefined
};

Textbox.propTypes = propTypes;
Textbox.defaultProps = defaultProps;

export default Textbox;
