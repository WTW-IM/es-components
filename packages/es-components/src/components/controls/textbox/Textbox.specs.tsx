import React from 'react';
import { fireEvent, screen } from '@testing-library/react';
import { renderWithTheme } from '../../util/test-utils';

import Control from '../Control';
import Label from '../label/Label';
import Textbox, { TextboxProps } from './Textbox';

const TestTextbox = (props: TextboxProps) => (
  <Control>
    <Label htmlFor="test">Text</Label>
    <Textbox id="test" {...props} />
  </Control>
);

it('executes the handleOnChange function when text is changed', () => {
  const props = {
    onChange: jest.fn()
  };
  renderWithTheme(<TestTextbox {...props} />);

  fireEvent.change(screen.getByLabelText('Text'), {
    target: { value: '112' }
  });
  expect(props.onChange).toHaveBeenCalled();
});

it('executes handleOnBlur when input focus is lost', () => {
  const props = {
    onBlur: jest.fn()
  };
  renderWithTheme(<TestTextbox {...props} />);

  fireEvent.blur(screen.getByLabelText('Text'));
  expect(props.onBlur).toHaveBeenCalled();
});

test.each<TextboxProps>([
  { prependIconName: 'phone', appendIconName: undefined },
  { prependIconName: undefined, appendIconName: 'user' },
  { prependIconName: 'phone', appendIconName: 'user' }
])('renders the correct addons when provided', props => {
  const { container } = renderWithTheme(<TestTextbox {...props} />);

  expect(container).toMatchSnapshot();
});
