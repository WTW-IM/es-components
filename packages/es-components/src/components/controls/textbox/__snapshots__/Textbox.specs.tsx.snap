// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct addons when provided 1`] = `
.c0 {
  color: #444;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 25px;
}

.c0 >[role="group"] {
  margin-bottom: 0;
}

.c1 {
  color: #444;
  cursor: pointer;
  font-family: 'Source Sans Pro','Segoe UI',Segoe,Calibri,Tahoma,sans-serif;
  font-size: 18px;
  font-weight: bold;
  line-height: 1.428;
  margin-bottom: 5px;
  display: inline-block;
}

.c1[disabled] {
  cursor: not-allowed;
}

.c3 {
  font-style: normal;
  font-weight: normal;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  display: inline-block;
  font-size: 18px;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: text-bottom;
}

.c4 {
  border: 1px solid #c5c5c5;
  border-radius: 2px;
  box-sizing: border-box;
  background-color: #fff;
  box-shadow: 0 0 0 0 #c5c5c5,inset 0 1px 1px rgba(0,0,0,0.075);
  -webkit-transition: border-color linear 0.15s,box-shadow linear 0.15s;
  transition: border-color linear 0.15s,box-shadow linear 0.15s;
  color: #000;
  font-size: 18px;
  font-family: "Source Sans Pro","Segoe UI",Segoe,Calibri,Tahoma,sans-serif;
  font-weight: normal;
  height: 2.2em;
  line-height: 1.428;
  min-width: 0;
  padding: 6px 12px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c4.c4 {
  outline: 0;
}

.c4::-webkit-input-placeholder {
  color: #787878;
}

.c4::-moz-placeholder {
  color: #787878;
}

.c4:-ms-input-placeholder {
  color: #787878;
}

.c4::placeholder {
  color: #787878;
}

.c4:disabled,
.c4:disabled:-moz-read-only {
  background-color: #ededed;
  cursor: text;
  cursor: not-allowed;
}

.c4:disabled,
.c4:disabled:read-only {
  background-color: #ededed;
  cursor: text;
  cursor: not-allowed;
}

.c2 {
  background-color: #fff;
  box-shadow: 0 0 0 0 #c5c5c5,inset 0 1px 1px rgba(0,0,0,0.075);
  -webkit-transition: border-color linear 0.15s,box-shadow linear 0.15s;
  transition: border-color linear 0.15s,box-shadow linear 0.15s;
  border: 1px solid #c5c5c5;
  border-radius: 2px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 !important;
}

.c2:focus-within {
  border-color: #83bffc;
  box-shadow: 0 0 8px rgba(102,175,233,0.6),inset 0 0 0 0 rgba(0,0,0,0);
}

.c2.c2:before {
  font-style: normal;
  font-weight: normal;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  display: inline-block;
  font-size: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: text-bottom;
  font-size: 18px;
  background-color: #e4e4e4;
  color: #5e5e5e;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  margin: 0;
  outline: 0;
  padding: 0.333em 0.6111em;
}

.c2.c2:after {
  font-style: normal;
  font-weight: normal;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  display: inline-block;
  font-size: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: text-bottom;
  font-size: 18px;
  background-color: #e4e4e4;
  color: #5e5e5e;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  margin: 0;
  outline: 0;
  padding: 0.333em 0.6111em;
}

.c5 {
  background-color: #fff;
  box-shadow: 0 0 0 0 #c5c5c5,inset 0 1px 1px rgba(0,0,0,0.075);
  -webkit-transition: border-color linear 0.15s,box-shadow linear 0.15s;
  transition: border-color linear 0.15s,box-shadow linear 0.15s;
  display: table-cell;
  -webkit-appearance: none;
  width: 100%;
  border-bottom-left-radius: 0;
  border-top-left-radius: 0;
}

.c5.c5:-moz-read-only {
  background-color: #ededed;
  cursor: text;
}

.c5.c5:read-only {
  background-color: #ededed;
  cursor: text;
}

.c5.c5 {
  border: none !important;
  margin: 0 !important;
  height: 2.2em;
}

.c5.c5:focus {
  box-shadow: inset 0 0 0 0 rgba(0,0,0,0);
}

@media (min-width:768px) {
  .c1 {
    margin-right: 20px;
  }
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
      for="test"
    >
      Text
    </label>
    <div
      class="c2"
    >
      <div
        aria-hidden="true"
        style="position: absolute; visibility: hidden; overflow: visible; width: 0px; height: 0px; left: -5000px;"
      >
        <input
          type="hidden"
        />
        <i
          aria-hidden="true"
          class="c3 bds-icon bds-phone"
        />
        <input
          type="hidden"
        />
        <i
          aria-hidden="true"
          class="c3 bds-icon bds-"
        />
      </div>
      <input
        class="c4 c5"
        id="test"
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`renders the correct addons when provided 2`] = `
.c0 {
  color: #444;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 25px;
}

.c0 >[role="group"] {
  margin-bottom: 0;
}

.c1 {
  color: #444;
  cursor: pointer;
  font-family: 'Source Sans Pro','Segoe UI',Segoe,Calibri,Tahoma,sans-serif;
  font-size: 18px;
  font-weight: bold;
  line-height: 1.428;
  margin-bottom: 5px;
  display: inline-block;
}

.c1[disabled] {
  cursor: not-allowed;
}

.c3 {
  font-style: normal;
  font-weight: normal;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  display: inline-block;
  font-size: 18px;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: text-bottom;
}

.c4 {
  border: 1px solid #c5c5c5;
  border-radius: 2px;
  box-sizing: border-box;
  background-color: #fff;
  box-shadow: 0 0 0 0 #c5c5c5,inset 0 1px 1px rgba(0,0,0,0.075);
  -webkit-transition: border-color linear 0.15s,box-shadow linear 0.15s;
  transition: border-color linear 0.15s,box-shadow linear 0.15s;
  color: #000;
  font-size: 18px;
  font-family: "Source Sans Pro","Segoe UI",Segoe,Calibri,Tahoma,sans-serif;
  font-weight: normal;
  height: 2.2em;
  line-height: 1.428;
  min-width: 0;
  padding: 6px 12px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c4.c4 {
  outline: 0;
}

.c4::-webkit-input-placeholder {
  color: #787878;
}

.c4::-moz-placeholder {
  color: #787878;
}

.c4:-ms-input-placeholder {
  color: #787878;
}

.c4::placeholder {
  color: #787878;
}

.c4:disabled,
.c4:disabled:-moz-read-only {
  background-color: #ededed;
  cursor: text;
  cursor: not-allowed;
}

.c4:disabled,
.c4:disabled:read-only {
  background-color: #ededed;
  cursor: text;
  cursor: not-allowed;
}

.c2 {
  background-color: #fff;
  box-shadow: 0 0 0 0 #c5c5c5,inset 0 1px 1px rgba(0,0,0,0.075);
  -webkit-transition: border-color linear 0.15s,box-shadow linear 0.15s;
  transition: border-color linear 0.15s,box-shadow linear 0.15s;
  border: 1px solid #c5c5c5;
  border-radius: 2px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 !important;
}

.c2:focus-within {
  border-color: #83bffc;
  box-shadow: 0 0 8px rgba(102,175,233,0.6),inset 0 0 0 0 rgba(0,0,0,0);
}

.c2.c2:before {
  font-style: normal;
  font-weight: normal;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  display: inline-block;
  font-size: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: text-bottom;
  font-size: 18px;
  background-color: #e4e4e4;
  color: #5e5e5e;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  margin: 0;
  outline: 0;
  padding: 0.333em 0.6111em;
}

.c2.c2:after {
  font-style: normal;
  font-weight: normal;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  display: inline-block;
  font-size: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: text-bottom;
  font-size: 18px;
  background-color: #e4e4e4;
  color: #5e5e5e;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  margin: 0;
  outline: 0;
  padding: 0.333em 0.6111em;
}

.c5 {
  background-color: #fff;
  box-shadow: 0 0 0 0 #c5c5c5,inset 0 1px 1px rgba(0,0,0,0.075);
  -webkit-transition: border-color linear 0.15s,box-shadow linear 0.15s;
  transition: border-color linear 0.15s,box-shadow linear 0.15s;
  display: table-cell;
  -webkit-appearance: none;
  width: 100%;
  border-bottom-right-radius: 0;
  border-top-right-radius: 0;
}

.c5.c5:-moz-read-only {
  background-color: #ededed;
  cursor: text;
}

.c5.c5:read-only {
  background-color: #ededed;
  cursor: text;
}

.c5.c5 {
  border: none !important;
  margin: 0 !important;
  height: 2.2em;
}

.c5.c5:focus {
  box-shadow: inset 0 0 0 0 rgba(0,0,0,0);
}

@media (min-width:768px) {
  .c1 {
    margin-right: 20px;
  }
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
      for="test"
    >
      Text
    </label>
    <div
      class="c2"
    >
      <div
        aria-hidden="true"
        style="position: absolute; visibility: hidden; overflow: visible; width: 0px; height: 0px; left: -5000px;"
      >
        <input
          type="hidden"
        />
        <i
          aria-hidden="true"
          class="c3 bds-icon bds-"
        />
        <input
          type="hidden"
        />
        <i
          aria-hidden="true"
          class="c3 bds-icon bds-user"
        />
      </div>
      <input
        class="c4 c5"
        id="test"
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`renders the correct addons when provided 3`] = `
.c0 {
  color: #444;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 25px;
}

.c0 >[role="group"] {
  margin-bottom: 0;
}

.c1 {
  color: #444;
  cursor: pointer;
  font-family: 'Source Sans Pro','Segoe UI',Segoe,Calibri,Tahoma,sans-serif;
  font-size: 18px;
  font-weight: bold;
  line-height: 1.428;
  margin-bottom: 5px;
  display: inline-block;
}

.c1[disabled] {
  cursor: not-allowed;
}

.c3 {
  font-style: normal;
  font-weight: normal;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  display: inline-block;
  font-size: 18px;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: text-bottom;
}

.c4 {
  border: 1px solid #c5c5c5;
  border-radius: 2px;
  box-sizing: border-box;
  background-color: #fff;
  box-shadow: 0 0 0 0 #c5c5c5,inset 0 1px 1px rgba(0,0,0,0.075);
  -webkit-transition: border-color linear 0.15s,box-shadow linear 0.15s;
  transition: border-color linear 0.15s,box-shadow linear 0.15s;
  color: #000;
  font-size: 18px;
  font-family: "Source Sans Pro","Segoe UI",Segoe,Calibri,Tahoma,sans-serif;
  font-weight: normal;
  height: 2.2em;
  line-height: 1.428;
  min-width: 0;
  padding: 6px 12px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c4.c4 {
  outline: 0;
}

.c4::-webkit-input-placeholder {
  color: #787878;
}

.c4::-moz-placeholder {
  color: #787878;
}

.c4:-ms-input-placeholder {
  color: #787878;
}

.c4::placeholder {
  color: #787878;
}

.c4:disabled,
.c4:disabled:-moz-read-only {
  background-color: #ededed;
  cursor: text;
  cursor: not-allowed;
}

.c4:disabled,
.c4:disabled:read-only {
  background-color: #ededed;
  cursor: text;
  cursor: not-allowed;
}

.c2 {
  background-color: #fff;
  box-shadow: 0 0 0 0 #c5c5c5,inset 0 1px 1px rgba(0,0,0,0.075);
  -webkit-transition: border-color linear 0.15s,box-shadow linear 0.15s;
  transition: border-color linear 0.15s,box-shadow linear 0.15s;
  border: 1px solid #c5c5c5;
  border-radius: 2px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 !important;
}

.c2:focus-within {
  border-color: #83bffc;
  box-shadow: 0 0 8px rgba(102,175,233,0.6),inset 0 0 0 0 rgba(0,0,0,0);
}

.c2.c2:before {
  font-style: normal;
  font-weight: normal;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  display: inline-block;
  font-size: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: text-bottom;
  font-size: 18px;
  background-color: #e4e4e4;
  color: #5e5e5e;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  margin: 0;
  outline: 0;
  padding: 0.333em 0.6111em;
}

.c2.c2:after {
  font-style: normal;
  font-weight: normal;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  display: inline-block;
  font-size: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: text-bottom;
  font-size: 18px;
  background-color: #e4e4e4;
  color: #5e5e5e;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  margin: 0;
  outline: 0;
  padding: 0.333em 0.6111em;
}

.c5 {
  background-color: #fff;
  box-shadow: 0 0 0 0 #c5c5c5,inset 0 1px 1px rgba(0,0,0,0.075);
  -webkit-transition: border-color linear 0.15s,box-shadow linear 0.15s;
  transition: border-color linear 0.15s,box-shadow linear 0.15s;
  display: table-cell;
  -webkit-appearance: none;
  width: 100%;
  border-bottom-left-radius: 0;
  border-top-left-radius: 0;
  border-bottom-right-radius: 0;
  border-top-right-radius: 0;
}

.c5.c5:-moz-read-only {
  background-color: #ededed;
  cursor: text;
}

.c5.c5:read-only {
  background-color: #ededed;
  cursor: text;
}

.c5.c5 {
  border: none !important;
  margin: 0 !important;
  height: 2.2em;
}

.c5.c5:focus {
  box-shadow: inset 0 0 0 0 rgba(0,0,0,0);
}

@media (min-width:768px) {
  .c1 {
    margin-right: 20px;
  }
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
      for="test"
    >
      Text
    </label>
    <div
      class="c2"
    >
      <div
        aria-hidden="true"
        style="position: absolute; visibility: hidden; overflow: visible; width: 0px; height: 0px; left: -5000px;"
      >
        <input
          type="hidden"
        />
        <i
          aria-hidden="true"
          class="c3 bds-icon bds-phone"
        />
        <input
          type="hidden"
        />
        <i
          aria-hidden="true"
          class="c3 bds-icon bds-user"
        />
      </div>
      <input
        class="c4 c5"
        id="test"
        type="text"
      />
    </div>
  </div>
</div>
`;
